diff --git a/src/terminal.c b/src/terminal.c
index d030f6d70..20fe73348 100644
--- a/src/terminal.c
+++ b/src/terminal.c
@@ -2751,6 +2752,21 @@ handle_pushline(int cols, const VTermScreenCell *cells, void *user)
 	    vim_free(((sb_line_T *)term->tl_scrollback.ga_data + i)->sb_cells);
 	    ml_delete(1, FALSE);
 	}
+	{
+	    const tabpage_T* tp;
+	    win_T* wp;
+	    FOR_ALL_TAB_WINDOWS(tp, wp)
+	    {
+		if (wp->w_buffer->b_term == term) {
+		    win_T* const orgWindowPtr = curwin;
+		    curwin = wp;
+		    check_cursor();
+		    curwin = orgWindowPtr;
+		    goto target_window_found;
+		}
+	    }
+	}
+target_window_found:
 	curbuf = curwin->w_buffer;
 
 	term->tl_scrollback.ga_len -= todo;
